keps/sig-api-machinery/4355-coordinated-leader-election/README.md

```diff
@@ -0,0 +1,966 @@
+- [ ] (R) Production readiness review completed
+- [ ] (R) Production readiness review approved
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+This proposes a component leader election mechanism that is safer for upgrades
+and rollbacks.
+
+This leader election approach continues to use leases, but with two
+key modifications:
+
+- Instead of a race by component instances to claim the lease, component instances
+  declare candidacy for a lease and a election coordinator claims the lease for the
+  best available candidate. This allows the election coordinator to pick a
+  candidate with the lowest version to ensure that skew rules are not violated.
+- The election coordinator can mark a lease as "end of term" to signal to the
+  current leader to stop renewing the lease. This allows the election
+  coordinator to preempt the current leader and replace it with a better one.
+
+## Motivation
+
+The most common upgrade approach used for Kubernetes control plane components is
+a node-by-node approach where all the component of a control plane node are
+terminated together and then restarted at the new version. This process is
+performed node-by-node across a high availability configuration. 
+
+Systems using node-by-node upgrades:
+
+- Cluster API
+- kubeadm
+- KIND
+
+To respect the [Kubernetes skew policy](https://kubernetes.io/releases/version-skew-policy/):
+
+- Upgrades should keep controller managers and schedulers at the *old* version until all apiservers
+  are upgraded.
+- Rollbacks should rollback controller managers and schedulers at the *old* version before any
+  apiservers are rolledback.
+
+But a node-by-node upgrade or rollback does not achieve this today.
+
+- For upgrade, there is about a 25% chance of a new version of the controller
+  running while old versions of the apiserver are active, resulting in a skew
+  violation. (Consider case where the 2nd node upgraded has the lease)
+- For rollback, it is almost a certainty that skew will be violated.
+
+There is also the possiblity that the lease will be lost by a leader during an
+upgrade or rollback resulting in the version of the controller flip-flopping
+between old and new.
+
+### Goals
```

Comment:
I like this KEP because it also facilitates other potential optimizations like graceful and performant failover, which a component can utilize to warm up (e.g. start informers to pull k8s objects) before actually becoming the leader. I don't want to make this KEP more complex than it needs to be but if you think there is anything that can be covered for future extensions, it might be great.

Reply:
This is a good point.  I was hoping this would be a good building block for potential future work around leadership transitions and controller scaling (breaking controllers into smaller groups or sharding controllers..).  I'll add some of this to the future work section